File: mycuration/src/main/java/com/phicdy/mycuration/presenter/FeedUrlHookPresenter.java
package com.phicdy.mycuration.presenter;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.phicdy.mycuration.db.DatabaseAdapter;
import com.phicdy.mycuration.rss.Feed;
import com.phicdy.mycuration.rss.RssParseExecutor;
import com.phicdy.mycuration.rss.RssParseResult;
import com.phicdy.mycuration.rss.RssParser;
import com.phicdy.mycuration.rss.UnreadCountManager;
import com.phicdy.mycuration.task.NetworkTaskManager;
import com.phicdy.mycuration.util.UrlUtil;
import com.phicdy.mycuration.view.FeedUrlHookView;

public class FeedUrlHookPresenter implements Presenter {
    private FeedUrlHookView view;
    private final DatabaseAdapter dbAdapter;
    private final UnreadCountManager unreadCountManager;
    private final NetworkTaskManager networkTaskManager;
    private final RssParser parser;
    RssParseExecutor.RssParseCallback callback = new RssParseExecutor.RssParseCallback() {
        @Override
        public void succeeded(@NonNull String url) {
            Feed newFeed = dbAdapter.getFeedByUrl(url);
            unreadCountManager.addFeed(newFeed);
            networkTaskManager.updateFeed(newFeed);
        }

        @Override
        public void failed(@RssParseResult.FailedReason int reason) {
            if (reason == RssParseResult.INVALID_URL) {
                view.showInvalidUrlErrorToast();
            } else {
                view.showGenericErrorToast();
            }
            view.dismissProgressDialog();
            view.finishView();
        }
    };

    public FeedUrlHookPresenter(@NonNull DatabaseAdapter dbAdapter,
                                @NonNull UnreadCountManager unreadCountManager,
                                @NonNull NetworkTaskManager networkTaskManager,
                                @NonNull RssParser parser) {
        this.dbAdapter = dbAdapter;
        this.unreadCountManager = unreadCountManager;
        this.networkTaskManager = networkTaskManager;
        this.parser = parser;
    }
    public void setView(@NonNull FeedUrlHookView view) {
        this.view = view;
    }

    @Override
    public void create() {
    }

    @Override
    public void resume() {
    }

    @Override
    public void pause() {
        view.unregisterFinishAddReceiver();
    }

    public void handle(@NonNull String action, @NonNull String url) {
        if (action.equals(Intent.ACTION_VIEW) || action.equals(Intent.ACTION_SEND)) {
            if (UrlUtil.isCorrectUrl(url)) {
                view.registerFinishAddReceiver();
                view.showProgressDialog();
                RssParseExecutor executor = new RssParseExecutor(parser, dbAdapter);
                executor.start(url, callback);
            }else {
                view.showInvalidUrlErrorToast();
            }
        }else {
            view.finishView();
        }
    }

    public void handleFinish(@NonNull String action, @Nullable String feedUrl,
                      int errorReason) {
        if (action.equals(NetworkTaskManager.FINISH_UPDATE_ACTION)) {
            view.dismissProgressDialog();
            view.showSuccessToast();
            view.finishView();
        }
    }
}

File: feature_feed_url_hook/src/main/java/com/phicdy/mycuration/feedurlhook/FeedUrlHookPresenter.kt
package com.phicdy.mycuration.feedurlhook

import android.content.Intent
import com.phicdy.mycuration.data.repository.RssRepository
import com.phicdy.mycuration.domain.rss.RssParseExecutor
import com.phicdy.mycuration.domain.rss.RssParseResult
import com.phicdy.mycuration.domain.rss.RssParser
import com.phicdy.mycuration.domain.rss.RssUrlHookIntentData
import com.phicdy.mycuration.domain.task.NetworkTaskManager
import com.phicdy.mycuration.util.UrlUtil
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import javax.inject.Inject

class FeedUrlHookPresenter @Inject constructor(
    private val view: FeedUrlHookView,
    private val rssUrlHookIntentData: RssUrlHookIntentData,
    private val rssRepository: RssRepository,
    private val networkTaskManager: NetworkTaskManager,
    private val coroutineScope: CoroutineScope,
    private val parser: RssParser
) {

    var callback: RssParseExecutor.RssParseCallback = object : RssParseExecutor.RssParseCallback {
        override fun succeeded(rssUrl: String) {
            coroutineScope.launch {
                val newFeed = rssRepository.getFeedByUrl(rssUrl)
                newFeed?.let {
                    networkTaskManager.updateFeed(newFeed)
                }
                view.showSuccessToast()
                view.finishView()
            }
        }

        override fun failed(reason: RssParseResult.FailedReason, url: String) {
            if (reason === RssParseResult.FailedReason.INVALID_URL) {
                view.showInvalidUrlErrorToast()
            } else {
                view.showGenericErrorToast()
            }
            view.trackFailedUrl(url)
            view.finishView()
        }
    }

    suspend fun create() {
        if (rssUrlHookIntentData.action != Intent.ACTION_VIEW && rssUrlHookIntentData.action != Intent.ACTION_SEND) {
            view.finishView()
            return
        }
        var url: String? = null
        if (rssUrlHookIntentData.action == Intent.ACTION_VIEW) {
            url = rssUrlHookIntentData.dataString
        } else if (rssUrlHookIntentData.action == Intent.ACTION_SEND) {
            // For Chrome
            url = rssUrlHookIntentData.extrasText.toString()
        }
        if (url != null) {
            handle(rssUrlHookIntentData.action, url)
        }
    }

    private suspend fun handle(action: String, url: String) {
        if (action == Intent.ACTION_VIEW || action == Intent.ACTION_SEND) {
            if (UrlUtil.isCorrectUrl(url)) {
                val executor = RssParseExecutor(parser, rssRepository)
                executor.start(url, callback)
            } else {
                view.showInvalidUrlErrorToast()
                view.trackFailedUrl(url)
            }
        } else {
            view.finishView()
        }
    }
}
